### ISSUE TITLE
[swift2objc] Add Extensions

### ISSUE THREAD

**Issue by nikeokoronkwo** [dart-lang/native#1815]
Extensions are a swift language feature that allows you to add methods and other nested declarations to a swift nestable declaration (struct or class). For exporting to Objective C, only classes are allowed to be extended (denoted by `@interface <class-name> (SWIFT_EXTENSION(<module-name>))` in Objective C)

```swift

```

They work similar to the relationship between Dart classes

Swift functionality present in an extension usually have a `swiftExtension` field in the generated symbol graph

**Comment by Hassnaa9** [comment]
Hi @liamappelbe  While implementing support for extensions, I considered two approaches and wanted to share my thinking and get your feedback.
Approach 1: Merge directly into the target CompoundDeclaration, after parsing all symbols, a `_mergeExtensionMembers` post processing step finds every `swift.extension symbol`, follows the extensionOf relation to the target type, and attaches the extension's methods, properties, and initializers directly to the already parsed CompoundDeclaration.
Approach 2: Parse extensions as a separate ExtensionDeclaration AST node like class and enum by create a new ExtensionDeclaration class and a parseExtensionDeclaration function. Then merge them into their target compounds in a second step.
My preference is Approach 1, because `ExtensionDeclaration` would never appear in the output, it always gets dissolved into its target type immediately after parsing. So Approach 2 adds a new AST node, a new parser file, and extra code paths to maintain, all for zero benefit over Approach 1 since the end result is identical as far as i know.

Also the only known limitation i think is extensions on external types (e.g. extension String ), where the target isn't in the module's symbolgraph, these are currently skipped with a warning. Should I open a separate issue to track that?
I would like to ensure that I clearly understand and if approach 1 will be compatible with the existing architecture before finalizing the PR. Thanks!

**Comment by liamappelbe** [comment]
@Hassnaa9 We tried approach 1 in FFIgen at first, and ran into issues such as the limitation you mentioned. As far as I know, extensions can be annotated with `@objc`, so the best approach would be to create an `ExtensionDeclaration` node, transform it like the other compounds, and codegen it as an `@objc extension`

**Comment by Hassnaa9** [comment]
> [@Hassnaa9](https://github.com/Hassnaa9) We tried approach 1 in FFIgen at first, and ran into issues such as the limitation you mentioned. As far as I know, extensions can be annotated with `@objc`, so the best approach would be to create an `ExtensionDeclaration` node, transform it like the other compounds, and codegen it as an `@objc extension`

Ok i will submit a PR with approach 2 solution

**Comment by liamappelbe** [comment]
Well, the approach I was suggesting wasn't the number 2 you listed. `ExtensionDeclaration` *does* appear in the output, as an `@objc extension`.

**Comment by Hassnaa9** [comment]
@liamappelbe Ok, another question please, after investigating the symbol graph, I found that there is no dedicated `swift.extension` symbol kind(unlike when we resolved the support operator overloading bug which has `swift.func.op`). Extension members appear as regular `swift.method`, `swift.property`, and `swift.init` symbols with a `swiftExtension` field marking them as coming from an extension. So i think the parser file isn't needed, instead, we identify extension members during parsing via the `swiftExtension` field, then use a post-processing step to group them into `ExtensionDeclaration` nodes, is this a good approach?

**Comment by liamappelbe** [comment]
Yeah, I think that's probably the best approach. When you see a method/property/init with this flag, instead of adding it directly to the target node, get-or-create an extension instead.

**Comment by Hassnaa9** [comment]
> Yeah, I think that's probably the best approach. When you see a method/property/init with this flag, instead of adding it directly to the target node, get-or-create an extension instead.

@liamappelbe Should I address support extensions on external types on this PR or report an issue with to extend it later? and put TODO comment for now

**Comment by liamappelbe** [comment]
I think if you use the `@objc extension` approach, it might "just work". If not, then yeah, just put a TODO and we can file a bug to do it later.

**Comment by Hassnaa9** [comment]
> I think if you use the `@objc extension` approach, it might "just work". If not, then yeah, just put a TODO and we can file a bug to do it later.

@liamappelbe Yeah i use `@objc extension` approach and doesn't work for external types i think this because they're parsed as BuiltInDeclaration and never go through parseCompoundDeclaration, so there's no compound to attach the extension to?

**Comment by liamappelbe** [comment]
Ok, just leave it as a TODO. We like to attach all our TODOs to bugs. You can probably find other examples in the repo for the syntax we use
